tests/docker/test-helpers.go

```diff
@@ -0,0 +1,480 @@
+	fmt.Println("Kubeconfig file: ", config.KubeconfigFile)
+	return nil
+}
+
+// RunCmdOnDocker runs a command on a docker container
+func RunCmdOnDocker(container, cmd string) (string, error) {
+	dCmd := fmt.Sprintf("docker exec %s %s", container, cmd)
+	return RunCommand(dCmd)
+}
+
+// RunCommand Runs command on the host.
+// Returns stdout and embeds stderr inside the error message.
+func RunCommand(cmd string) (string, error) {
+	var stdout, stderr bytes.Buffer
+	c := exec.Command("bash", "-c", cmd)
+	c.Stdout = &stdout
+	c.Stderr = &stderr
+	err := c.Run()
+	if err != nil {
+		return stdout.String(), fmt.Errorf("failed to run command: %s: %s: %v", cmd, stderr.String(), err)
+	}
+	return stdout.String(), nil
+}
+
+func checkVersionSkew(config *TestConfig) error {
+	if config.NumAgents > 0 {
+		serverImage := getEnvOrDefault("K3S_IMAGE_SERVER", config.K3sImage)
+		agentImage := getEnvOrDefault("K3S_IMAGE_AGENT", config.K3sImage)
+		if semver.Compare(semver.MajorMinor(agentImage), semver.MajorMinor(serverImage)) > 0 {
+			return fmt.Errorf("agent version cannot be higher than server - not supported by Kubernetes version skew policy")
+		}
+	}
+
+	return nil
+}
+
+func getEnvOrDefault(key, defaultValue string) string {
+	if value := os.Getenv(key); value != "" {
+		return value
+	}
+	return defaultValue
+}
+
+// VerifyValidVersion checks for invalid version strings
+func VerifyValidVersion(container string, binary string) error {
+	output, err := RunCmdOnDocker(container, binary+" version")
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(output, "\n")
+	// Check for invalid version strings
+	re := regexp.MustCompile(`(?i).*(dev|head|unknown|fail|refuse|\+[^"]*\.).*`)
+	for _, line := range lines {
+		if re.MatchString(line) {
+			return fmt.Errorf("invalid version string found in %s: %s", binary, line)
+		}
+	}
+
+	return nil
+}
+
+// Returns the latest version from the update channel
+func GetVersionFromChannel(upgradeChannel string) (string, error) {
+	url := fmt.Sprintf("https://update.k3s.io/v1-release/channels/%s", upgradeChannel)
+	client := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	resp, err := client.Get(url)
+	if err != nil {
+		return "", fmt.Errorf("failed to get URL: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusFound {
+		return "", fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	finalURL := resp.Header.Get("Location")
+	if finalURL == "" {
+		return "", fmt.Errorf("location header not set")
+	}
+	version := finalURL[strings.LastIndex(finalURL, "/")+1:]
+	version = strings.Replace(version, "+", "-", 1)
+	return version, nil
+}
+
+// TODO the below functions are duplicated in the integration test utils. Consider combining into commmon package
+// DeploymentsReady checks if the provided list of deployments are ready, otherwise returns an error
```

Comment:
👍🏻 on combining these as soon as possible.

We also have another copy in rke2. Maybe we need a common repo for the test framework. Maybe there already is one from k8s and we should be using that instead.